File: ina/code/iii.py
import time
import random
import networkx as nx
from collections import deque

def isolated(G, i):
  for j in G[i]:
    if j != i:
      return False
  return True
  
def component(G, N, i):
  C = []
  S = []
  N.remove(i)
  S.append(i)
  while S:
    i = S.pop()
    C.append(i)
    for j in G[i]:
      if j in N:
        N.remove(j)
        S.append(j)
  return C

def components(G):
  C = []
  N = set(G.nodes())
  while N:
    C.append(component(G, N, next(iter(N))))
  return C
  
def distance(G, i):
  D = [-1] * len(G) # D = {}
  Q = deque()
  D[i] = 0
  Q.append(i)
  while Q:
    i = Q.popleft()
    for j in G[i]:
      if D[j] == -1: # if j not in D:
        D[j] = D[i] + 1
        Q.append(j)
  return [d for d in D if d > 0]
  
def distances(G, n = 100):
  D = [] # D = {}
  for i in G.nodes() if len(G) <= n else random.sample(G.nodes(), n):
    D.append(distance(G, i)) # D[i] = distance(G, i)
  return D
  
def erdos_renyi(n, m):
  G = nx.MultiGraph(name = "erdos_renyi")
  for i in range(n):
    G.add_node(i)
  edges = []
  while len(edges) < m:
    i = random.randint(0, n - 1)
    j = random.randint(0, n - 1)
    if i != j:
      edges.append((i, j))
  G.add_edges_from(edges)
  return G
  
def configuration(degrees):
  G = nx.MultiGraph(name = "configuration")
  nodes = []
  for i in range(len(degrees)):
    G.add_node(i)
    for _ in range(degrees[i]):
      nodes.append(i)
  random.shuffle(nodes)
  edges = []
  for i in range(0, len(nodes), 2):
    edges.append((nodes[i], nodes[i + 1]))
  G.add_edges_from(edges)
  return G
  
def info(G):
  print("{:>10s} | '{:s}'".format('Graph', G.name))
  
  tic = time.time()
  n = G.number_of_nodes()
  n0, n1, delta = 0, 0, 0
  for i in G.nodes():
    if isolated(G, i):
      n0 += 1
    elif G.degree(i) == 1:
      n1 += 1
    if G.degree(i) > delta:
      delta = G.degree(i)
  
  print("{:>10s} | {:,d} ({:,d}, {:,d})".format('Nodes', n, n0, n1))
  
  m = G.number_of_edges()
  m0 = nx.number_of_selfloops(G)
  
  print("{:>10s} | {:,d} ({:,d})".format('Edges', m, m0))
  print("{:>10s} | {:.2f} ({:,d})".format('Degree', 2 * m / n, delta))
  print("{:>10s} | {:.2e}".format('Density', 2 * m / n / (n - 1)))
      
  C = components(G)

  print("{:>10s} | {:.1f}% ({:,d})".format('LCC', 100 * max(len(c) for c in C) / n, len(C)))
  
  D = distances(G, 100)
  D = [i for d in D for i in d]

  print("{:>10s} | {:.2f} ({:,d})".format('Distance', sum(D) / len(D), max(D)))
  
  if isinstance(G, nx.MultiGraph):
    G = nx.Graph(G)
    
  print("{:>10s} | {:.4f}".format('Clustering', nx.average_clustering(G)))
  
  print("{:>10s} | {:.1f} s".format('Time', time.time() - tic))
  print()

for name in ["toy", "karate_club", "collaboration_imdb", "www_google"]:
  # G = nx.Graph(nx.read_pajek("/Users/lovre/Downloads/" + name + ".net"))
  # G.name = name
  
  G = nx.Graph(name = name)
  with open("/Users/lovre/Downloads/" + name + ".net", 'r') as file:
    file.readline()
    
    for line in file:
      if line.startswith("*"):
        break
      else:
        node = line.split("\"")
        G.add_node(int(node[0]) - 1, label = node[1])
        
    for line in file:
      i, j = (int(x) - 1 for x in line.split()[:2])
      G.add_edge(i, j)
  
  info(G)
  
  info(erdos_renyi(G.number_of_nodes(), G.number_of_edges()))
  
  info(configuration([k for _, k in G.degree()]))

File: ina/code/III.java
package si.lj.uni.fri.lna.test.ina.labs;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

import si.lj.uni.fri.lna.LNA;
import si.lj.uni.fri.lna.entity.graph.Graph;
import si.lj.uni.fri.lna.entity.graph.ListGraph;
import si.lj.uni.fri.lna.io.graph.Readers;
import si.lj.uni.fri.lna.logic.utility.Information;
import si.lj.uni.fri.lna.test.bib.Complexity;

public class III {

	public static void main(String[] args) throws Exception {
		LNA.init();
		
		for (String file: new String[] { "toy", "karate_club", "collaboration_imdb", "www_google" }) {
			Complexity.pic();
			
			Graph graph = Readers.pajek("/Users/lovre/Downloads/" + file + ".net");
			
			info(graph);
			
			info(random(graph));
			
			Complexity.poc();
		}
		
		LNA.exit();
	}
	
	public static void info(Graph graph) throws Exception {
		int n0 = 0, n1 = 0, delta = 0;
		for (int i = 0; i < graph.getN(); i++) {
			if (graph.getDegree(i) == 0)
				n0++;
			else if (graph.getDegree(i) == 1)
				n1++;
			if (graph.getDegree(i) > delta)
				delta = graph.getDegree(i);
		}
		
		List<List<Integer>> Cs = components(graph);
		int S = 0;
		for (List<Integer> C: Cs)
			if (C.size() > S)
				S = C.size();
		
		int diameter = 0;
		double distance = Double.NaN;
		if (graph.getN() < 100000) {
			int[][] Ds = distances(graph);
			distance = 0.0; diameter = 0;
			for (int i = 0; i < graph.getN(); i++)
				for (int j = 0; j < graph.getN(); j++) {
					distance += 1.0 * Ds[i][j] / graph.getN() / graph.getN();
					if (Ds[i][j] > diameter)
						diameter = Ds[i][j];
				}
		}
		
		Information.println(String.format("\n%12s | '%s'", "Graph", graph.getName()));
		Information.println(String.format("%12s | %,d (%,d, %,d)", "Nodes", graph.getN(), n0, n1));
		Information.println(String.format("%12s | %,d", "Edges", graph.getM()));
		Information.println(String.format("%12s | %.3f (%,d)", "Degree", graph.getDegree(), delta));
		Information.println(String.format("%12s | %.3e", "Density", graph.getDensity()));
		Information.println(String.format("%12s | %.3f (%,d)", "Distance", distance, diameter));
		Information.println(String.format("%12s | %.1f%% (%,d)", "LCC", 100.0 * S / graph.getN(), Cs.size()));
	}
	
	public static List<List<Integer>> components(Graph graph) {
		Set<Integer> N = new HashSet<Integer>();
		for (int i = 0; i < graph.getN(); i++)
			N.add(i);
		
		List<List<Integer>> Cs = new LinkedList<List<Integer>>();
		while (!N.isEmpty())
			Cs.add(component(graph, N, N.iterator().next()));
		
		return Cs;
	}
	
	public static List<Integer> component(Graph graph, Set<Integer> N, int i) {
		List<Integer> C = new LinkedList<Integer>();
		List<Integer> S = new LinkedList<Integer>();
		
		S.add(i);
		N.remove(i);
		
		while (!S.isEmpty()) {
			i = S.remove(0);
			C.add(i);
			
			for (int j: graph.getNeighbors(i))
				if (N.remove(j))
					S.add(0, j);
		}
		
		return C;
	}
	
	public static int[][] distances(Graph graph) {
		int[][] Ds = new int[graph.getN()][];
		for (int i = 0; i < graph.getN(); i++)
			Ds[i] = distance(graph, i);
		
		return Ds;
	}
	
	public static int[] distance(Graph graph, int i) {
		int[] D = new int[graph.getN()];
		for (int j = 0; j < graph.getN(); j++)
			D[j] = -1;
		List<Integer> Q = new LinkedList<Integer>();
		
		D[i] = 0;
		Q.add(i);
		
		while (!Q.isEmpty()) {
			i = Q.remove(0);
			
			for (int j: graph.getNeighbors(i))
				if (D[j] == -1) {
					D[j] = D[i] + 1;
					Q.add(j);
				}
		}
		
		for (int j = 0; j < graph.getN(); j++)
			if (D[j] == -1)
				D[j] = 0;
		
		return D;
	}
	
	public static Graph random(Graph graph) {
		Graph random = new ListGraph("erdos_renyi");
		
		for (int i = 0; i < graph.getN(); i++)
			random.addNode();
		
		while (random.getM() < graph.getM()) {
			int i = (int)(Math.random() * random.getN());
			int j = (int)(Math.random() * random.getN());
			
			if (i != j)
				random.addEdge(i, j);
		}
		
		return random;
	}

}
